File: KodiInterop/KodiInterop/Modules/Xbmc/XbmcMonitor.cs
using Smx.KodiInterop.Messages;
using Smx.KodiInterop.Python;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace Smx.KodiInterop.Modules.Xbmc
{
    public class XbmcMonitor : IDisposable, IKodiEventConsumer
    {
		public readonly PyVariable Instance;
		
		public event EventHandler<EventArgs> OnAbortRequested;
		public event EventHandler<LibraryEventArgs> OnCleanStarted;
		public event EventHandler<LibraryEventArgs> OnCleanFinished;
		public event EventHandler<DatabaseEventArgs> OnDatabaseScanStarted;
		public event EventHandler<DatabaseEventArgs> OnDatabaseUpdated;
		public event EventHandler<EventArgs> OnDPMSActivated;
		public event EventHandler<EventArgs> OnDPMSDeactivated;
		public event EventHandler<EventArgs> OnScreensaverActivated;
		public event EventHandler<EventArgs> OnScreensaverDeactivated;
		public event EventHandler<LibraryEventArgs> OnScanStarted;
		public event EventHandler<LibraryEventArgs> OnScanFinished;
		public event EventHandler<EventArgs> OnSettingsChanged;
		public event EventHandler<NotificationEventArgs> OnNotification;

		public bool TriggerEvent(KodiEventMessage e) {
			switch (e.Source) {
				case "onAbortRequested":
					OnAbortRequested?.Invoke(null, new EventArgs());
					break;
				case "onCleanStarted":
					OnCleanStarted?.Invoke(null, new LibraryEventArgs(e.EventArgs[0]));
					break;
				case "onCleanFinished":
					OnCleanFinished?.Invoke(null, new LibraryEventArgs(e.EventArgs[0]));
					break;
				case "onDPMSActivated":
					OnDPMSActivated?.Invoke(null, new EventArgs());
					break;
				case "onDPMSDeactivated":
					OnDPMSDeactivated?.Invoke(null, new EventArgs());
					break;
				case "onDatabaseScanStarted":
					OnDatabaseScanStarted?.Invoke(null, new DatabaseEventArgs(e.EventArgs[0]));
					break;
				case "onDatabaseUpdated":
					OnDatabaseUpdated?.Invoke(null, new DatabaseEventArgs(e.EventArgs[0]));
					break;
				case "onNotification":
					OnNotification?.Invoke(null, new NotificationEventArgs(
						e.EventArgs[0], e.EventArgs[1], e.EventArgs[2]
					));
					break;
				case "onScanStarted":
					OnScanStarted?.Invoke(null, new LibraryEventArgs(e.EventArgs[0]));
					break;
				case "onScanFinished":
					OnScanFinished?.Invoke(null, new LibraryEventArgs(e.EventArgs[0]));
					break;
				case "onScreensaverActivated":
					OnScreensaverActivated?.Invoke(null, new EventArgs());
					break;
				case "onScreensaverDeactivated":
					OnScreensaverDeactivated?.Invoke(null, new EventArgs());
					break;
				case "onSettingsChanged":
					OnSettingsChanged?.Invoke(null, new EventArgs());
					break;
				default:
					PyConsole.WriteLine(string.Format("Unknown event '{0}' not handled", e.Source));
					return false;
			}
			return true;
		}

		public void Dispose() {
			KodiBridge
				.RunningAddon
				.Bridge
				.UnregisterEventClass(this);
			Instance.Dispose();
		}

		public bool AbortRequested {
			get => Convert.ToBoolean(Instance.CallFunction("abortRequested"));
		}

		public bool WaitForAbort(TimeSpan timeSpan) {
			return Convert.ToBoolean(Instance.CallFunction("waitForAbort", new List<object> {
				(ulong)timeSpan.TotalSeconds
			}));
		}

		public XbmcMonitor() {
            this.Instance = PyVariableManager.Get.NewVariable(evalCode: "self.new_monitor()");

			// We now register this type so that PostEvent will be able to invoke onMessage in this class
			Console.WriteLine("=> Registering EventClass " + typeof(XbmcMonitor).FullName);
			KodiBridge
				.RunningAddon
				.Bridge
				.RegisterMonitor(this);
		}
	}
}

File: csharp/xbmc_monitor.py
import xbmc
import json

class xbmc_monitor(xbmc.Monitor):

	def __init__(self):
		super(xbmc_monitor, self).__init__()

	def set_bridge(self, bridge):
		self.bridge = bridge

	def post(self, eventName, *args, **kwargs):
		jDict = {
			"sender": "Monitor",
			"source": eventName,
			"args": args,
			"kwargs": kwargs
		}
		self.bridge.post_event(json.dumps(jDict))

	def onAbortRequested(self, *args, **kwargs):
		self.post("onAbortRequested", *args, **kwargs)

	def onCleanStarted(self, *args, **kwargs):
		self.post("onCleanStarted", *args, **kwargs)
	
	def onCleanFinished(self, *args, **kwargs):
		self.post("onCleanFinished", *args, **kwargs)
	
	def onDPMSActivated(self, *args, **kwargs):
		self.post("onDPMSActivated", *args, **kwargs)
	
	def onDPMSDeactivated(self, *args, **kwargs):
		self.post("onDPMSDeactivated", *args, **kwargs)
	
	def onDatabaseScanStarted(self, *args, **kwargs):
		self.post("onDatabaseScanStarted", *args, **kwargs)
	
	def onDatabaseUpdated(self, *args, **kwargs):
		self.post("onDatabaseUpdated", *args, **kwargs)
	
	def onNotification(self, *args, **kwargs):
		self.post("onNotification", *args, **kwargs)
	
	def onScanStarted(self, *args, **kwargs):
		self.post("onScanStarted", *args, **kwargs)
	
	def onScanFinished(self, *args, **kwargs):
		self.post("onScanFinished", *args, **kwargs)
	
	def onScreensaverActivated(self, *args, **kwargs):
		self.post("onScreensaverActivated", *args, **kwargs)
	
	def onScreensaverDeactivated(self, *args, **kwargs):
		self.post("onScreensaverDeactivated", *args, **kwargs)
	
	def onSettingsChanged(self, *args, **kwargs):
		self.post("onSettingsChanged", *args, **kwargs)
